### PR TITLE
8231111: Cgroups v2: Rework Metrics in java.base so as to recognize unified hierarchy

### PR DESCRIPTION
**Pull request by zhengyu123**
This is a followup backport for Cgroup v2 support.

The original patch does not apply cleanly, due to context differences and some of out-of-order backports, conflicts were resolved manually.

In additional:

1) Include partial backport of JDK-8250627.
    - Renamed `Metrics.c` =>  `CgroupMetrics.c`
    - Added native method `isUseContainerSupport()` to `CgroupMetrics.java`
    - Return instance only if `isUseContainerSupport()` is enabled.
   
2) Add `createTempDirectory()` method to `test/lib/jdk/test/lib/Utils.java` for `TestCgroupSubsystemController.java` test

3) Add `import jdk.test.lib.process.OutputAnalyzer;` to `TestDockerMemoryMetrics.java`

Test:

- [x] jtreg containers/docker test on Ubuntu 20.04.4 LTS
- [x] jdk/internal/platform/cgroup on Ubuntu 20.04.4 LTS
- [x] jdk/internal/platform/docker on Ubuntu 20.04.4 LTS

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8231111](https://bugs.openjdk.java.net/browse/JDK-8231111): Cgroups v2: Rework Metrics in java.base so as to recognize unified hierarchy

### Reviewers
 * [Severin Gehwolf](https://openjdk.java.net/census#sgehwolf) (@jerboaa - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk11u-dev pull/863/head:pull/863` \
`$ git checkout pull/863`

Update a local copy of the PR: \
`$ git checkout pull/863` \
`$ git pull https://git.openjdk.java.net/jdk11u-dev pull/863/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 863`

View PR using the GUI difftool: \
`$ git pr show -t 863`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk11u-dev/pull/863.diff">https://git.openjdk.java.net/jdk11u-dev/pull/863.diff</a>

</details>

### CODE REVIEW

File: src/java.base/linux/classes/jdk/internal/platform/cgroupv1/CgroupV1Subsystem.java [hunk 1]
```diff
@@ -0,0 +1,490 @@
...
+         *
+         * For a Host from memory example above the path would be:
+         *
+         * /sys/fs/cgroup/memory/user.slice
+         *
+         */
+        try (Stream<String> lines =
+                CgroupUtil.readFilePrivileged(Paths.get("/proc/self/cgroup"))) {
+
+            lines.map(line -> line.split(":"))
```
Comment (jerboaa) [inline src/java.base/linux/classes/jdk/internal/platform/cgroupv1/CgroupV1Subsystem.java:108]: We are missing the patch of JDK-8272124 here. Old code was:
https://github.com/openjdk/jdk11u-dev/blob/194b66f6a72437345c458e747e24beb46352537a/src/java.base/linux/classes/jdk/internal/platform/cgroupv1/Metrics.java#L119..L122

Best to use the JDK 17u commit as a base for this:
See: https://github.com/openjdk/jdk17u/commit/f0028333d83c39b434c26f66a6a92e446e7047cd#diff-7d6f202c58b4205f19a9b6c657f4827e9ef132d7433a7bf906b5b640f9f7841fR197

Looking at this again, we should bring in the test changes of JDK-8272124 too (as we are now bringing in those tests with the cgroups v2 backport). It helps to verify this fix is present as well.

File: src/java.base/linux/classes/jdk/internal/platform/cgroupv1/CgroupV1Subsystem.java [hunk 2]
```diff
@@ -0,0 +1,490 @@
...
+     */
+    private static void setSubSystemControllerPath(CgroupV1Subsystem subsystem, String[] entry) {
+        String controllerName;
+        String base;
+        CgroupV1SubsystemController controller = null;
+        CgroupV1SubsystemController controller2 = null;
+
+        controllerName = entry[1];
+        base = entry[2];
+        if (controllerName != null && base != null) {
```
Comment (jerboaa) [inline src/java.base/linux/classes/jdk/internal/platform/cgroupv1/CgroupV1Subsystem.java:168]: We are missing the fix of https://bugs.openjdk.java.net/browse/JDK-8254854 here. Old code was:
https://github.com/openjdk/jdk11u-dev/blob/194b66f6a72437345c458e747e24beb46352537a/src/java.base/linux/classes/jdk/internal/platform/cgroupv1/Metrics.java#L200..L202

See head commit at the time:
https://github.com/openjdk/jdk/commit/a0b687bf

File: src/java.base/linux/classes/jdk/internal/platform/cgroupv1/CgroupV1Subsystem.java [hunk 3]
```diff
@@ -0,0 +1,490 @@
...
+        }
+    }
+
+
+    private static boolean getHierarchical(CgroupV1MemorySubSystemController controller) {
+        long hierarchical = getLongValue(controller, "memory.use_hierarchy");
+        return hierarchical > 0;
+    }
+
+    private void setActiveSubSystems() {
```
Comment (jerboaa) [inline src/java.base/linux/classes/jdk/internal/platform/cgroupv1/CgroupV1Subsystem.java:216]: function `getSwapEnabled()` from JDK-8250984 is missing here. See:
https://github.com/openjdk/jdk11u-dev/blob/194b66f6a72437345c458e747e24beb46352537a/src/java.base/linux/classes/jdk/internal/platform/cgroupv1/Metrics.java#L246..L249

File: src/java.base/linux/classes/jdk/internal/platform/cgroupv1/CgroupV1Subsystem.java [hunk 4]
```diff
@@ -0,0 +1,490 @@
...
+        return getLongValue(memory, "memory.kmem.tcp.max_usage_in_bytes");
+    }
+
+    public long getTcpMemoryUsage() {
+        return getLongValue(memory, "memory.kmem.tcp.usage_in_bytes");
+    }
+
+    public long getMemoryAndSwapFailCount() {
+        return getLongValue(memory, "memory.memsw.failcnt");
+    }
```
Comment (jerboaa) [inline src/java.base/linux/classes/jdk/internal/platform/cgroupv1/CgroupV1Subsystem.java:442]: This is missing the hunk for https://bugs.openjdk.java.net/browse/JDK-8250984. See:
https://github.com/openjdk/jdk11u-dev/blob/87703322e980e42035d7cf101f04cafdc31fe278/src/java.base/linux/classes/jdk/internal/platform/cgroupv1/Metrics.java#L480..L482

File: src/java.base/linux/classes/jdk/internal/platform/cgroupv1/CgroupV1Subsystem.java [hunk 5]
```diff
@@ -0,0 +1,490 @@
...
+
+    public long getMemoryAndSwapLimit() {
+        long retval = getLongValue(memory, "memory.memsw.limit_in_bytes");
```
Comment (jerboaa) [inline src/java.base/linux/classes/jdk/internal/platform/cgroupv1/CgroupV1Subsystem.java:445]: This is missing the hunk for https://bugs.openjdk.java.net/browse/JDK-8250984. See:
https://github.com/openjdk/jdk11u-dev/blob/87703322e980e42035d7cf101f04cafdc31fe278/src/java.base/linux/classes/jdk/internal/platform/cgroupv1/Metrics.java#L487..L489

File: src/java.base/linux/classes/jdk/internal/platform/cgroupv1/CgroupV1Subsystem.java [hunk 6]
```diff
@@ -0,0 +1,490 @@
...
+                retval = CgroupV1SubsystemController.getLongValueMatchingLine(memory,
+                                                            "memory.stat",
+                                                            match);
+            }
+        }
+        return CgroupV1SubsystemController.longValOrUnlimited(retval);
+    }
+
+    public long getMemoryAndSwapMaxUsage() {
+        return getLongValue(memory, "memory.memsw.max_usage_in_bytes");
```
Comment (jerboaa) [inline src/java.base/linux/classes/jdk/internal/platform/cgroupv1/CgroupV1Subsystem.java:460]: This is missing the hunk for https://bugs.openjdk.java.net/browse/JDK-8250984. See:
https://github.com/openjdk/jdk11u-dev/blob/87703322e980e42035d7cf101f04cafdc31fe278/src/java.base/linux/classes/jdk/internal/platform/cgroupv1/Metrics.java#L506..L508

File: src/java.base/linux/classes/jdk/internal/platform/cgroupv1/CgroupV1Subsystem.java [hunk 7]
```diff
@@ -0,0 +1,490 @@
...
+    }
+
+    public long getMemoryAndSwapUsage() {
+        return getLongValue(memory, "memory.memsw.usage_in_bytes");
```
Comment (jerboaa) [inline src/java.base/linux/classes/jdk/internal/platform/cgroupv1/CgroupV1Subsystem.java:464]: This is missing the hunk for https://bugs.openjdk.java.net/browse/JDK-8250984. See:
https://github.com/openjdk/jdk11u-dev/blob/87703322e980e42035d7cf101f04cafdc31fe278/src/java.base/linux/classes/jdk/internal/platform/cgroupv1/Metrics.java#L513..L515

**Review by jerboaa — CHANGES_REQUESTED** [review]
A few more comments related to: https://bugs.openjdk.java.net/browse/JDK-8250984

**Comment by zhengyu123** [comment]
> 

@jerboaa Thanks for the review. I underestimated how many out-of-order backports that affected this backport, especially files without conflicts. I redid `CgroupV1Subsystem.java` and need to go through others.

Thanks.

**Comment by jerboaa** [comment]
> @jerboaa Thanks for the review. I underestimated how many out-of-order backports that affected this backport, especially files without conflicts. I redid `CgroupV1Subsystem.java` and need to go through others.

Anytime. Please let me know when you are ready for another review.

**Comment by zhengyu123** [comment]
> getMemoryAndSwapMaxUsage

Ready for another round, thanks.

-Zhengyu

**Review by jerboaa — CHANGES_REQUESTED** [review]
A few more fix-ups are needed (see also inline comments):

1. [JDK-8272124](https://bugs.openjdk.java.net/browse/JDK-8272124) fix and test.
2. [JDK-8223396](https://bugs.openjdk.java.net/browse/JDK-8223396). Backport those test changes as a dependency?
3. `PlainRead.java` would need to get the `-2` added as a "good" value. Like here: https://github.com/openjdk/jdk/commit/01875677#diff-fb4f4dbf8943e1dcfd1663922c16f9ed3db403c070bd50979680ec5303756674R48 It's part of the original JDK-8250984 fix, but wasn't part of the 11 backport when it was done. So we should add it with this backport.

**Comment by jerboaa** [comment]
@zhengyu123 Please also merge latest master into your branch so that GHA tests become clean (baring build/test fail due to this patch).

**Comment by zhengyu123** [comment]
> A few more fix-ups are needed (see also inline comments):
> > 3. `PlainRead.java` would need to get the `-2` added as a "good" value. Like here: [openjdk/jdk@01875677#diff-fb4f4dbf8943e1dcfd1663922c16f9ed3db403c070bd50979680ec5303756674R48](https://github.com/openjdk/jdk/commit/01875677#diff-fb4f4dbf8943e1dcfd1663922c16f9ed3db403c070bd50979680ec5303756674R48) It's part of the original JDK-8250984 fix, but wasn't part of the 11 backport when it was done. So we should add it with this backport.

 Even with this change, the test still failed.

**Comment by jerboaa** [comment]
> > A few more fix-ups are needed (see also inline comments):
> > > 3. `PlainRead.java` would need to get the `-2` added as a "good" value. Like here: [openjdk/jdk@01875677#diff-fb4f4dbf8943e1dcfd1663922c16f9ed3db403c070bd50979680ec5303756674R48](https://github.com/openjdk/jdk/commit/01875677#diff-fb4f4dbf8943e1dcfd1663922c16f9ed3db403c070bd50979680ec5303756674R48) It's part of the original JDK-8250984 fix, but wasn't part of the 11 backport when it was done. So we should add it with this backport.
> 
> Even with this change, the test still failed.

Right. See: https://github.com/openjdk/jdk11u-dev/pull/840#issuecomment-1067065313

**Comment by zhengyu123** [comment]
> > > A few more fix-ups are needed (see also inline comments):
> > > > 3. `PlainRead.java` would need to get the `-2` added as a "good" value. Like here: [openjdk/jdk@01875677#diff-fb4f4dbf8943e1dcfd1663922c16f9ed3db403c070bd50979680ec5303756674R48](https://github.com/openjdk/jdk/commit/01875677#diff-fb4f4dbf8943e1dcfd1663922c16f9ed3db403c070bd50979680ec5303756674R48) It's part of the original JDK-8250984 fix, but wasn't part of the 11 backport when it was done. So we should add it with this backport.
> > 
> > 
> > Even with this change, the test still failed.
> 
> Right. See: [#840 (comment)](https://github.com/openjdk/jdk11u-dev/pull/840#issuecomment-1067065313)

Okay. It does not fix the test, I don't think we should include it in this patch, I would prefer to fix it separately.

**Comment by jerboaa** [comment]
> > > > A few more fix-ups are needed (see also inline comments):
> > > > > 3. `PlainRead.java` would need to get the `-2` added as a "good" value. Like here: [openjdk/jdk@01875677#diff-fb4f4dbf8943e1dcfd1663922c16f9ed3db403c070bd50979680ec5303756674R48](https://github.com/openjdk/jdk/commit/01875677#diff-fb4f4dbf8943e1dcfd1663922c16f9ed3db403c070bd50979680ec5303756674R48) It's part of the original JDK-8250984 fix, but wasn't part of the 11 backport when it was done. So we should add it with this backport.
> > > 
> > > 
> > > Even with this change, the test still failed.
> > 
> > 
> > Right. See: [#840 (comment)](https://github.com/openjdk/jdk11u-dev/pull/840#issuecomment-1067065313)
> 
> Okay. It does not fix the test, I don't think we should include it in this patch, I would prefer to fix it separately.

@zhengyu123 That's fine with me. Please rebase on top of the test changes ([JDK-8223396](https://bugs.openjdk.java.net/browse/JDK-8223396)) integrated recently and backport the test and fix from [JDK-8272124](https://bugs.openjdk.java.net/browse/JDK-8272124). Thank you.

**Comment by jerboaa** [comment]
> > > > > > A few more fix-ups are needed (see also inline comments):
> > > > > > > 3. `PlainRead.java` would need to get the `-2` added as a "good" value. Like here: [openjdk/jdk@01875677#diff-fb4f4dbf8943e1dcfd1663922c16f9ed3db403c070bd50979680ec5303756674R48](https://github.com/openjdk/jdk/commit/01875677#diff-fb4f4dbf8943e1dcfd1663922c16f9ed3db403c070bd50979680ec5303756674R48) It's part of the original JDK-8250984 fix, but wasn't part of the 11 backport when it was done. So we should add it with this backport.
> > > > > 
> > > > > 
> > > > > Even with this change, the test still failed.
> > > > 
> > > > 
> > > > Right. See: [#840 (comment)](https://github.com/openjdk/jdk11u-dev/pull/840#issuecomment-1067065313)
> > > 
> > > 
> > > Okay. It does not fix the test, I don't think we should include it in this patch, I would prefer to fix it separately.
> > 
> > 
> > @zhengyu123 That's fine with me. Please rebase on top of the test changes ([JDK-8223396](https://bugs.openjdk.java.net/browse/JDK-8223396)) integrated recently and backport the test and fix from [JDK-8272124](https://bugs.openjdk.java.net/browse/JDK-8272124). Thank you.
> 
> I think all 4 PRs are lined up for push. Once this review is done, I will request for approvals for all 4.

Sorry my bad, I thought we were missing the fix from JDK-8272124 but it's there now. As for the test, we can only add that in once JDK-8239559 gets backported.

**Review by jerboaa — APPROVED** [review]
This looks good now. Thanks!

**Pull request closed without merging**